### PR TITLE
Fix cost accounting, GGUF resolution and teardown safety found in liv…

### LINKED ISSUE #6: Bugfixes
# Follow-ups from #2: GGUF default quantisation, billed-rate accounting, explicit .gguf refs

Found while verifying the multi-GPU work (#2, PR #5). Nothing here was introduced by that PR; 1 and 3 reproduce on `main`. Ordered by impact.

## 1. GGUF engines ignore the Q4_K_M default and pick full precision

**What happens.** Without `--quantization`, llama.cpp picks the repository's BF16 file, not Q4_K_M. The README and the cookbook both say GGUF engines default to Q4_K_M.

```
$ larri up --model unsloth/Qwen3-8B-GGUF --runtime llamacpp --dry-run
  weights     Qwen3-8B-BF16.gguf (15.3 GB)
    sizing     unsloth/Qwen3-8B-GGUF needs ~19.6 GB VRAM
```

That is about 3× the download, and a bigger card than the operator expected, rented without being asked.

**Cause.** In `cmd/larri/main.go`, `pickRuntime` runs before `spec.Quantization = quantFor(eng, *quant)`. For llama.cpp, `pickRuntime` calls `newLlamaCpp`, which calls `llamacpp.ResolveGGUF` right away with an empty quantisation. The default is applied after the file has already been picked.

**Fix.** The engine has to be chosen before the default can be known, so resolve the weights in a separate step after `quantFor`, not inside `newLlamaCpp`. Add a test: an empty `--quantization` on a GGUF repo resolves to the Q4_K_M file.

## 2. The journal costs a rig at the quoted rate, not the billed one

**What happens.** `larri status` shows the billed rate beside the quote, but the accrued cost is computed from the quote. When they differ, the cost shown is wrong, and usually too low:

```
hardware  vastai · 2× CMP 170HX 128GB · $0.821/hr (quoted $0.802)
hardware  vastai · RTX 3060 12GB · $0.051/hr (quoted $0.041)
```

On RunPod the gap was $2.78 quoted against $3.18 billed, about 14% low, until #2 made the catalogue quote the Secure tier. `internal/state` should never understate cost.

**Cause.** `Store.RecordIntent` always records `PriceHr: rig.Offer.PriceHr`. `state.Cost` takes the latest `PriceHr` from the journal, so the rate the provider reports after create never reaches the cost calculation.

**Fix.** Once an instance exists, journal `rig.BilledPriceHr()`. **Check first:** Vast's instance `dph_total` may already include storage, and `Cost` adds `StorageHr` separately. Switching the source as-is could double-count storage. Confirm what `dph_total` covers before deciding which rate goes where.

## 3. A ref naming a specific .gguf file fails at sizing

**What happens.**

```
$ larri up --model unsloth/Qwen3-8B-GGUF/Qwen3-8B-Q4_K_M.gguf --runtime llamacpp --dry-run
  weights     Qwen3-8B-Q4_K_M.gguf
    sizing     resolving unsloth/Qwen3-8B-GGUF/Qwen3-8B-Q4_K_M.gguf
larri: sizing: /api/models/unsloth/Qwen3-8B-GGUF/Qwen3-8B-Q4_K_M.gguf/revision/main not found
```

**Cause.** The runtime strips the filename with `llamacpp.RepoOf`. The Hugging Face facts lookup in sizing uses the whole ref as a repo id. The explicit-file path also returns `Weights{File: f}` with no size, so even when it works it sizes from the parameter-count estimate. The listing it could read the real size from is never fetched.

**Fix.** Sizing should use the repository part of the ref. For an explicit file, fetch the listing anyway and take that file's size, including all shards when it is part 1 of N.

## 4. RunPod lists a GPU type that it refuses to rent

**What happens.** `NVIDIA RTX PRO 4500 Blackwell Server Edition` appears in the RunPod catalogue with a price and stock. It passes `purchasable()`, but `POST /pods` rejects it because it is not in the create call's `gpuTypeIds` enum. When it is the cheapest fit, LARRI spends one create attempt on it. The failure is excluded and the run falls back, so nothing is billed, but it uses up one of the attempts.

**Cause.** `purchasable()` filters out only empty/unknown ids and MIG partitions. It deliberately avoids hard-coding the create enum, which goes stale.

**Fix (options).** Check offers against the create enum from a source that stays current (RunPod's published API schema, if one exposes it). Or add a small deny-list, with a conformance check that fails once the type becomes creatable.

## 5. A tight layer-split fit across three cards has not been tested on real hardware

**What happens.** Sizing accepts a three-card host with about 2% headroom per card:

```
$ larri up --model unsloth/Llama-3.3-70B-Instruct-GGUF --quantization Q6_K --runtime llamacpp \
    --gpu "RTX 3090" --gpus 3 --max-gpus 3 --dry-run
    sizing     needs ~67.3 GB VRAM
    sizing     sharding across all 3 cards
    sizing     per card: about 22.4 GB of 22.8 GB usable, 18.0 GB of it weights
```

The per-card figure assumes the model splits evenly. llama.cpp places whole layers, plus the output layer, on specific cards, so one card can end up holding more than the average. The only live attempt (3× RTX 3090) failed at the provider before the model loaded, so nobody knows whether this fit holds.

**Fix.** Verify it live. If it OOMs, either add extra headroom when a layer-split fit is tight, or pass an explicit per-card split to llama.cpp.

## 6. A record from before the fix still shows $111.17 for a pod that never existed

**What happens.** `larri status --all` shows a RunPod A100 SXM rig at $111.17 over 79h58m. Its create was refused on 2026-09-08 because `RUNPOD_API_KEY` was missing, so no pod ever existed. The rig was destroyed on 2026-09-12, before PR #5 added the rule that zeroes cost when a teardown records that nothing was created. New failures resolve correctly; this old record can't be fixed after the fact.

**Why there is no automatic fix.** `larri down` refuses a rig that is already destroyed. A provider listing today also can't prove nothing existed four days ago, so re-checking would conclude something the evidence doesn't support.

**Options.** Leave it as history. Or add a way for the operator to annotate a terminated rig with evidence they checked themselves (for example, the provider's billing page), recorded as coming from the operator.

## 7. `sizing.IsGGUF` is dead code

Only `TestIsGGUFSelectsLlamaCpp` calls it. Its doc comment says it selects llama.cpp, but runtime selection doesn't use it. Remove it, or use it in `pickRuntime` if a k-quant name should imply llama.cpp.

### PR DESCRIPTION
**Pull request by ramkatru**
…e runs

- Resolve GGUF weights in the daemon after the quantisation default, so up, offers, TUI and MCP all get Q4_K_M, not BF16
- Let MCP and `larri tui --runtime llamacpp` use llama.cpp; they failed resolving an empty model ref
- Accept `owner/repo/path/file.gguf` refs: size by repository, measure, check against the listing, record the file's quantisation
- Serve a named GGUF file as `qwen3-8b-q8_0` rather than the file name or its first part
- Size a quantisation published twice as one download, so advice no longer offers a larger file as a saving
- Quote Vast offers for the disk actually rented (allocated_storage), so the quote matches the bill
- Read Vast storage_cost as $/GB/month, not $/hr; it added $0.20/hr to every rig
- Price RunPod container disk and volume beside the GPU rate in quotes and instance rates
- Journal the billed rate, count storage once, and mark entries with the rules they were written under
- Read old journal entries by their original rules, at the billed rate the rig file kept where available
- Write each rig's termination to the journal entry that ends it; the field existed but was never set
- Resolve failed creates: close never-sent ones, tear down labelled instances, keep unknown outcomes FAILED
- Stop cost only on recorded nothing_created evidence, never on a blank instance id
- Add `larri down --nothing-created "<how you checked>" <rig>` to correct records LARRI cannot settle
- Act on a rig only through the provider holding it; down, resume and MCP used the configured default
- Keep a destroyed rig destroyed: the store refuses revival and cost replay ignores entries after the end
- End supervision when another process destroys the rig, or when probes fail and the provider no longer has it
- Offer only RunPod GPU types its create schema accepts, which drops the PRO 4500 Server Edition
- Report llama.cpp download percentage while weights arrive
- Measure download progress in bytes written, not blocks the filesystem reserved
- Follow the symlinked Hugging Face cache on RunPod so vLLM progress no longer sits at 0%
- Remove unused sizing.IsGGUF and move its cases into the runtime-selection test
- Add a regression test that the SSH host-key scan never completes a handshake
- Update the requirements spec, PROJECT_STATE, design doc, CLAUDE.md, README and site pages

### CODE REVIEW

File: internal/runtime/llamacpp/llamacpp.go [hunk 1]
```diff
@@ -142,14 +164,98 @@
...
+				continue
+			}
+			last = got
+			pct := float64(got) / float64(total) * 100
+			if pct > 100 {
+				pct = 100
+			}
+			send(runtime.Progress{
+				Phase: "weights.download", Percent: pct,
+				BytesDone: int64(got), BytesTotal: int64(total),
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/runtime/llamacpp/llamacpp.go]: Fixed in commit cb84e92.

File: internal/daemon/up.go [hunk 2]
```diff
@@ -1346,6 +1474,14 @@ func (o *Orchestrator) reportExclusions(sel rank.Result) {
 // motive afterwards gets it wrong exactly when several conditions were true at
 // once (§13.1).
 func (o *Orchestrator) Down(ctx context.Context, rig *core.Rig, term *core.Termination) error {
+	if err := o.holdsRig(rig, "daemon.Down"); err != nil {
+		return err
+	}
+	// Already over, and the record says why. Teardown is idempotent
+	// (FR-DEL-02), and a second one would journal a second ending.
+	if rig.State == core.StateDestroyed {
+		return nil
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/daemon/up.go:1485]: Fixed. Down re-reads the store before doing anything and adopts an ending another process already recorded. If it only finds out when it writes, the store refuses a stale second DESTROYED (ErrAlreadyDestroyed) and Down adopts the first record. Test: TestTwoTeardownsRecordOneEnding.

File: internal/daemon/up.go [hunk 3]
```diff
@@ -1361,8 +1497,28 @@ func (o *Orchestrator) Down(ctx context.Context, rig *core.Rig, term *core.Termi
...
+			if checked {
+				if term.Evidence == nil {
+					term.Evidence = map[string]string{}
+				}
+				term.Evidence[core.EvidenceNothingCreated] = "no instance carries this rig's label"
+			} else {
+				o.warn("destroy", "could not reach the provider to confirm nothing was created")
+			}
+			rig.End = term
+			return o.Store.Transition(rig, core.StateDestroyed, "no instance was ever created")
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/daemon/up.go]: Fixed. When the check by label can't be answered, Down returns a transient error and leaves the rig billable. It only records DESTROYED with nothing_created evidence. Test: TestDownOfARigWithNoInstanceWaitsForTheProvider.

File: internal/provider/runpod/types.go [hunk 4]
```diff
@@ -419,6 +443,13 @@ func (p pod) normalise() (core.Instance, error) {
 	if p.AdjustedCostPerHr != nil && *p.AdjustedCostPerHr > 0 {
 		inst.PriceHr = *p.AdjustedCostPerHr
 	}
+	// costPerHr is the GPU alone. The disks bill beside it, so the rate the
+	// pod costs while it runs is both, and what it costs stopped is the
+	// volume at the stopped rate.
+	if inst.PriceHr > 0 {
+		inst.PriceHr += runningDiskHr(p.ContainerDiskInGb, p.VolumeInGb)
+	}
+	inst.StorageHr = stoppedDiskHr(p.VolumeInGb)
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/provider/runpod/types.go:452]: Fixed in replay. A running hour costs PriceHr, all counted as compute, and a stopped hour costs StorageHr, matching CostSummary.StorageUSD ("accrued while STOPPED"). There's no subtraction any more, so a large RunPod volume can't overstate. Test: TestRunningAndStoppedRatesAreSeparate.

File: internal/runtime/llamacpp/gguf.go [hunk 5]
```diff
@@ -125,11 +107,46 @@ func ResolveGGUF(ctx context.Context, ref, quant string, token secret.Secret) (W
...
+// draft head, or a quantisation other than the one the rig would be recorded
+// and sized as.
+func namedFile(repo, file, quant string, ggufs []string, sizes map[string]uint64) (Weights, error) {
+	// llama.cpp is pointed at part one and finds the rest itself, so naming
+	// any part names the set.
+	first := ShardFiles(file)[0]
+	if _, ok := sizes[first]; !ok {
+		return Weights{}, errs.Newf(errs.ClassModelFailure, "llamacpp.ResolveGGUF",
+			"%s has no file %s; it carries: %s", repo, file, strings.Join(quantsIn(ggufs), ", "))
+	}
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/runtime/llamacpp/gguf.go]: Fixed. The named file must be in the listing, and completeSet requires every part, for named refs and for listing picks alike. A missing part is refused before renting. Test: TestAnIncompleteSetIsRefusedBeforeRenting.

File: internal/state/cost.go [hunk 6]
```diff
@@ -40,7 +64,16 @@ func Cost(entries []Entry, now time.Time) core.CostSummary {
...
+			// PriceHr is everything billed while the machine runs, storage
+			// included, so compute is what is left once storage is counted
+			// on its own line. Adding the two double-counted storage. An
+			// entry from before that rule held the quote, with storage on
+			// top.
+			if legacy && billedHr == 0 {
+				sum.ComputeUSD += hours * priceHr
+			} else {
+				sum.ComputeUSD += hours * math.Max(priceHr-storageHr, 0)
+			}
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/state/cost.go]: Fixed with the RunPod thread: running and stopped rates are accrued separately.

File: internal/runtime/llamacpp/llamacpp.go [hunk 7]
```diff
@@ -55,6 +56,27 @@ func (r *Runtime) SetWeights(w Weights) { r.weights = w }
...
+// The default quantisation is applied here, before the listing is read,
+// because this is the only place that knows it is needed. A ref naming a file
+// gets no default: the file already says what it carries, and a default
+// would contradict any file that is not Q4_K_M.
+func (r *Runtime) ResolveWeights(ctx context.Context, spec core.ModelSpec) (runtime.Weights, error) {
+	quant := spec.Quantization
+	if quant == "" && explicitFile(spec.Ref) == "" {
+		quant = r.DefaultQuantization()
+	}
+	w, err := ResolveGGUF(ctx, spec.Ref, quant, r.hfToken)
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/runtime/llamacpp/llamacpp.go]: Fixed. The revision now reaches the listing (/api/models/<repo>/revision/<rev>) and the advice lookup. Test: TestResolutionReadsTheRevisionItWillDownload.

File: internal/state/store.go [hunk 8]
```diff
@@ -95,11 +100,28 @@ func (s *Store) RecordIntent(rig *core.Rig, to core.LifecycleState, note string)
...
 // Transition journals a completed move and updates the snapshot.
 func (s *Store) Transition(rig *core.Rig, to core.LifecycleState, note string) error {
+	// A destroyed rig stays destroyed. Two processes can hold the same rig —
+	// one serving it, one running `larri down` — and the one holding a stale
+	// copy wrote DEGRADED over DESTROYED, which put a pod that no longer
+	// existed back to billing in the journal.
+	if to != core.StateDestroyed {
+		if cur, err := s.Load(rig.ID); err == nil && cur != nil && cur.State == core.StateDestroyed {
+			return fmt.Errorf("state: rig %s is already destroyed", rig.ID)
+		}
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/state/store.go]: Fixed. A DESTROYED write from a copy that isn't itself destroyed is refused with ErrAlreadyDestroyed and the first record stands. A destroyed copy can still record the nothing_created correction. Test: TestAStaleSecondTeardownKeepsTheFirstRecord.

[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/state/store.go]: Fixed. An unreadable rig file fails closed for every move except teardown, which a bad file must never block. Test: TestAnUnreadableRigFileFailsClosedExceptForTeardown.

File: internal/tools/larri.go [hunk 9]
```diff
@@ -26,7 +26,11 @@ type Deps struct {
 	// NewOrchestrator builds one configured for the current environment. It
 	// returns an error rather than a zero value so a missing API key is
 	// reported to the calling agent, not discovered as a nil dereference.
-	NewOrchestrator func(runtimeKind string) (*daemon.Orchestrator, error)
+	//
+	// providerName is the provider holding an existing rig, or "" for the
+	// configured default when there is no rig yet. Acting on a rig through
+	// any other provider reads "not found" as confirmed absence.
+	NewOrchestrator func(runtimeKind, providerName string) (*daemon.Orchestrator, error)
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/tools/larri.go]: Fixed. The factory now takes the model and chooses the engine from it as the CLI does, so named .gguf files and GGUF quantisations get llama.cpp. A bare *-GGUF repo with no quantisation still picks vLLM in both CLI and MCP; that's a separate runtime-selection gap. Test: TestTheEngineIsChosenFromTheModelTheAgentNamed.

File: internal/daemon/absence.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	if !checked {
+		o.warn("record", "could not reach %s to look for the rig's label; recording on your word alone",
+			rig.ProviderName())
+	}
+
+	end := core.Termination{Actor: core.ActorOperator, Code: core.ReasonOperatorRequest,
+		Summary: "recorded by the operator as never created"}
+	if rig.End != nil {
+		end = *rig.End
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/daemon/absence.go:80]: Fixed. The fallback ending is dated when the correction is made. Test: TestACorrectionToARigWithNoEndingIsDated.

File: internal/state/store.go [hunk 11]
```diff
@@ -95,11 +101,27 @@ func (s *Store) RecordIntent(rig *core.Rig, to core.LifecycleState, note string)
...
+// ErrAlreadyDestroyed is returned for a move from a stale copy of a rig that
+// the store already holds as destroyed. Nothing was written; the record that
+// ended it stands.
+var ErrAlreadyDestroyed = errors.New("rig already destroyed")
+
 // Transition journals a completed move and updates the snapshot.
 func (s *Store) Transition(rig *core.Rig, to core.LifecycleState, note string) error {
+	if err := s.guardDestroyed(rig, to); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/state/store.go:133]: Fixed. Save now passes the same destroyed-rig guard as Transition, and both check and write under a per-rig flock held across processes. Bring-up stops if its save is refused because the rig was ended elsewhere. Tests: TestAStaleSnapshotWriteCannotUndoAnEnding, TestTheDestroyedCheckRunsUnderTheWriteLock.

File: internal/state/store.go [hunk 12]
```diff
@@ -95,11 +101,27 @@ func (s *Store) RecordIntent(rig *core.Rig, to core.LifecycleState, note string)
 		e.Instance = rig.Instance.InstanceID
 		e.StorageHr = rig.Instance.StorageHr
 	}
+	// The entry that ends a rig says why. The field existed and nothing set
+	// it, so the reason lived only in the snapshot — and cost, which is
+	// replayed from the journal, never saw the evidence that a rig had
+	// created nothing: one confirmed absent at teardown was still costed
+	// at $2.09/hr for the minutes before it.
+	if to == core.StateDestroyed && rig.End != nil {
+		e.Termination = rig.End
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/state/store.go:111]: Fixed. When adopt finds the instance gone, it records a dated provider/instance-gone termination with the instance as evidence before the DESTROYED transition. Test: TestAdoptOfAVanishedInstanceRecordsWhy.

File: internal/tools/lifecycle.go [hunk 13]
```diff
@@ -326,7 +326,7 @@ func (d Deps) orphanDestroy(ctx context.Context, raw json.RawMessage) (any, erro
 	if a.InstanceID == "" {
 		return nil, fmt.Errorf("instance_id is required")
 	}
-	o, err := d.NewOrchestrator("")
+	o, err := d.NewOrchestrator("", "", core.ModelSpec{})
```
[the remark this answers is not in the document]
Comment (ramkatru) [reply internal/tools/lifecycle.go]: Fixed in two places. larri_orphan_destroy requires provider, which larri_orphans now returns per orphan. DestroyOrphan also confirms the instance is at that provider before destroying, and refuses otherwise, so no surface can have a wrong provider's "not found" read as absence. Tests: TestOrphansAreFoundAndDestroyedAtTheProviderHoldingThem, TestAnOrphanIsNeverDestroyedThroughAnotherProvider.

**Review by sivasundeep — APPROVED** [review]
